src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationProvider.cs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Extensions.Configuration
+{
+    /// <summary>
+    /// Base class for file based <see cref="ConfigurationProvider"/>.
+    /// </summary>
+    public abstract class FileConfigurationProvider : ConfigurationProvider
+    {
+        public FileConfigurationProvider(FileConfigurationSource source)
+        {
+            if (source == null)
+            {
+                throw new ArgumentNullException(nameof(source));
+            }
+            Source = source;
+        }
+
+        public FileConfigurationSource Source { get; }
+
+        public override void Initialize(IConfigurationRoot root)
+        {
+            if (root == null)
+            {
+                throw new ArgumentNullException(nameof(root));
+            }
+
+            // TODO: aggregate watch?
+            // REVIEW: should providers validate the source here?
+            if (Source.ReloadOnChange && Source.FileProvider != null)
+            {
+                ChangeToken.OnChange(
+                    () => Source.FileProvider.Watch(Source.Path),
+                    () =>
+                    {
+                        Load();
+                        root.RaiseChanged();
```

Comment:
That would work but seems a bit complicated.  Maybe we should just introduce some kind of IChangeTracker to encapsulate the ability to fire changes more than once, and pass the root's tracker to the providers.  We just need something that's more persistent since the change token only works for one firing today...

We could do it scoped to only config or add it primitives and make it general purpose.

```
   public interface IConfigurationChangeTracker {
      public void RegisterOnChanged(Action<IConfigurationRoot> callback)
      public void RaiseChanged();
   }

   IConfigurationRoot would add a ChangeTracker readonly property
   
```